src/dynamics/mpas/dyn_coupling.F90

```diff
@@ -0,0 +1,650 @@
+    use qneg, only: qneg_run
+    use static_energy, only: update_dry_static_energy_run
+    use string_utils, only: stringify
+
+    ! Modules from CESM Share.
+    use shr_kind_mod, only: kind_cx => shr_kind_cx, kind_r8 => shr_kind_r8
+
+    implicit none
+
+    private
+    ! Provide APIs required by CAM-SIMA.
+    public :: dynamics_to_physics_coupling
+    public :: physics_to_dynamics_coupling
+contains
+    !> Perform one-way coupling from the dynamics output states to the physics input states.
+    !> The other coupling direction is implemented by its counterpart, `physics_to_dynamics_coupling`.
+    !> (KCW, 2024-07-31)
+    subroutine dynamics_to_physics_coupling()
+        character(*), parameter :: subname = 'dyn_coupling::dynamics_to_physics_coupling'
+        integer :: column_index
+        integer, allocatable :: is_water_species_index(:)
+        integer, pointer :: index_qv
+        ! Variable name suffixes have the following meanings:
+        ! `*_col`: Variable is of each column.
+        ! `*_int`: Variable is at layer interfaces.
+        ! `*_mid`: Variable is at layer midpoints.
+        real(kind_r8), allocatable :: pd_int_col(:), &       ! Dry hydrostatic air pressure (Pa).
+                                      pd_mid_col(:), &       ! Dry hydrostatic air pressure (Pa).
+                                      p_int_col(:), &        ! Full hydrostatic air pressure (Pa).
+                                      p_mid_col(:), &        ! Full hydrostatic air pressure (Pa).
+                                      z_int_col(:)           ! Geometric height (m).
+        real(kind_r8), allocatable :: dpd_col(:), &          ! Dry air pressure difference (Pa) between layer interfaces.
+                                      dp_col(:), &           ! Full air pressure difference (Pa) between layer interfaces.
+                                      dz_col(:)              ! Geometric height difference (m) between layer interfaces.
+        real(kind_r8), allocatable :: qv_mid_col(:), &       ! Water vapor mixing ratio (kg kg-1).
+                                      sigma_all_q_mid_col(:) ! Summation of all water species mixing ratios (kg kg-1).
+        real(kind_r8), allocatable :: rhod_mid_col(:), &     ! Dry air density (kg m-3).
+                                      rho_mid_col(:)         ! Full air density (kg m-3).
+        real(kind_r8), allocatable :: t_mid_col(:), &        ! Temperature (K).
+                                      tm_mid_col(:), &       ! Modified "moist" temperature (K).
+                                      tv_mid_col(:)          ! Virtual temperature (K).
+        real(kind_r8), allocatable :: u_mid_col(:), &        ! Eastward wind velocity (m s-1).
+                                      v_mid_col(:), &        ! Northward wind velocity (m s-1).
+                                      omega_mid_col(:)       ! Vertical wind velocity (Pa s-1).
+        real(kind_r8), pointer :: exner(:, :)
```

Comment:
For variables that point directly to memory allocated by the MPAS-A dycore, would it be better to declare them as `real(rkind)`? I suppose at present the MPAS-A definition of `RKIND` always agrees with the CAM-SIMA definition of `kind_r8`, but there may in future be interest in running the dycore in single precision.

Reply:
MPAS dycore in CAM-SIMA employs OOP design. From the perspective of CAM-SIMA, MPAS dycore looks like a single opaque object [[1]](https://github.com/ESCOMP/CAM-SIMA/blob/81331442873166f48a2d569934440afdac43d6fd/src/dynamics/mpas/dyn_comp.F90#L79). All MPAS-specific datatypes and interworkings are abstracted and hidden away from CAM-SIMA.

Here, the quoted code section belongs to the realm of CAM-SIMA. It does not know what `rkind` is because the Fortran module that defines `rkind` in MPAS is unavailable to it.

I am aware that from MPAS version 8.1.0 onwards, single precision build has become the default. I am still experimenting the best way to do mixed precision interoperability between CAM-SIMA and MPAS dycore. I propose that we wait until the finalization of MPAS dycore is in CAM-SIMA, so I can implement the feature of mixed precision interoperability in one go.

Reply:
I'm fine with deferring work to match the kind-type of these variables with their kind-type in the MPAS-A dycore until a later PR.